lib/veri.ml

```diff
@@ -194,66 +190,57 @@ class ['a] t arch dis =
-        1) among load events that occured at current step, in the same context,
-           with the same address;
-        2) among load events that occures at current step, in other context,
-           with the same address;
-        3) in current context. *)
-    method private resolve_addr ~mem ~addr endian size =
-      self#eval_exp addr >>= fun addr_res ->
-      match value addr_res with
-      | Bil.Bot | Bil.Mem _ -> SM.return addr_res
-      | Bil.Imm addr' ->
-        SM.get () >>= fun ctxt ->
-        match find_mem_load (self_events ctxt) (same_addr addr') with
-        | Some mv -> self#store_and_load ~mem ~addr mv endian size
-        | None ->
-          match find_mem_load (other_events ctxt) (same_addr addr') with
-          | None -> super#eval_load ~mem ~addr endian size
-          | Some mv -> self#store_and_load ~mem ~addr mv endian size
-
-    (** [eval_load ~mem ~addr endian size] - returns a result bound with [addr].
-        Search starts from self events, if it was write access to given
-        address at current step. And if it was, then result of write access
-        returned and no load event is emitted.
-        Otherwise searching continues as written above for [resolve_addr],
-        with emitting memory load event. *)
-    method! eval_load ~mem ~addr endian size =
-      SM.get () >>= fun ctxt ->
-      self#eval_exp addr >>= fun addr_res ->
-      match value addr_res with
-      | Bil.Bot | Bil.Mem _ -> SM.return addr_res
-      | Bil.Imm addr' ->
-        match find_mem_store (self_events ctxt) (same_addr addr') with
-        | Some mv -> self#eval_exp (Bil.int (Move.data mv))
-        | None ->
-          self#resolve_addr mem addr endian size >>= fun r ->
-          match value addr_res, value r with
-          | Bil.Imm addr, Bil.Imm data ->
-            self#update_event (create_mem_load addr data) >>= fun () ->
+  method! eval_store ~mem ~addr data endian size =
+    super#eval_store ~mem ~addr data endian size >>= fun r ->
+    self#eval_exp addr >>= fun addr ->
+    self#eval_exp data >>= fun data ->
+    match value addr, value data with
+    | Bil.Imm addr, Bil.Imm data ->
+      SM.update (fun c -> c#discard_event (is_previous_store addr)) >>= fun () ->
+      let ev = create_mem_store addr data in
+      self#update_event ev >>= fun () -> SM.return r
+    | _ -> SM.return r
+
+  (** [find_addr_data addr] - returns a move data, bound with [addr].
+      Sequence of searches is the following:
+      1) among store events that occured at current step, in the same context,
+         with the same address; we don't need to emit event in this case
+      2) among load events that occured at current step, in the same context,
+         with the same address; will emit memory_load event
+      3) among load events that occures at current step, in other context,
+         with the same address; will emit memory_load event *)
+  method private find_addr_data addr =
```

Comment:
I already saw the same code in the `find_var_data`, it looks like that you are missing an abstraction opportunity. 

Also, if you have a parameter of type `bool` then it usually should be labeled.